File: app/src/main/java/com/example/producttest/adapter/ProductAdapter.java
package com.example.producttest.adapter;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.producttest.activity.MainActivity;
import com.example.producttest.R;
import com.example.producttest.activity.ProductDetailActivity;
import com.example.producttest.database.DatabaseHelper;
import com.example.producttest.model.Product;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class ProductAdapter extends RecyclerView.Adapter<ProductAdapter.ViewHolderProduct> {
    Context context;
    List<Product> productList;
    private final  int PICK_IMAGE_REQUEST = 71;
    EditText edtName;
    EditText edtPrice;
    EditText edtDes;
    TextView textViewId;
    AlertDialog alertDialog;

    public ProductAdapter(Context context, List<Product> productList) {
        this.context = context;
        this.productList = productList;
    }

    @NonNull
    @Override
    public ViewHolderProduct onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_product,parent,false);
        return new ViewHolderProduct(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderProduct holder, int position) {
        Product product = productList.get(position);
//        holder.txtId.setText("Id "+product.getId());
        holder.txtName.setText(product.getName());
        Locale locale = new Locale("vi","VN");
        NumberFormat fmt =NumberFormat.getCurrencyInstance(locale);
        holder.txtPrice.setText(fmt.format(product.getPrice()));
        holder.txtDes.setText(product.getDes());
//
//        byte [] imageProduct = product.getImage();
//        Bitmap bitmap = BitmapFactory.decodeByteArray(imageProduct,0,imageProduct.length);
//        holder.imgProduct.setImageBitmap(bitmap);

        holder.imgDelete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                androidx.appcompat.app.AlertDialog.Builder builder = new androidx.appcompat.app.AlertDialog.Builder(context);
                builder.setTitle("Delete Cart");
                builder.setMessage("Do you want to delete: "+productList.get(position).getName()+"");
                builder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int positions) {
                        removeProduct(position);
                        ((MainActivity)context).setTotalTextView();
                    }
                });
                builder.setNegativeButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {

                    }
                });
                builder.show();



            }
        });
    }

    @Override
    public int getItemCount() {
        return productList.size();
    }


    private int totalProduct(){
        int total = 0;
        for(Product product : productList)
            total+=product.getPrice();
        Log.d("sd",total+"");
        return total;
    }
    public class ViewHolderProduct extends  RecyclerView.ViewHolder  {
        public TextView txtName,txtPrice,txtDes,txtId;
        public ImageView imgProduct,imgDelete;
        public ImageView img_Productss;

        public ViewHolderProduct(@NonNull View itemView) {
            super(itemView);
//            txtId = itemView.findViewById(R.id.txtId);
            img_Productss = itemView.findViewById(R.id.img_Productss);
            txtName = itemView.findViewById(R.id.txtName);
            txtPrice = itemView.findViewById(R.id.txtPrice);
            txtDes = itemView.findViewById(R.id.txtDes);
            imgProduct = itemView.findViewById(R.id.imgProduct);
            imgDelete = itemView.findViewById(R.id.delete_cart);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
//
                    Intent intent = new Intent(context, ProductDetailActivity.class);
                    intent.putExtra("ProductKey",productList.get(getPosition()));
                    context.startActivity(intent);
                }
            });
            ///
            itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    CharSequence [] item  = {"Update","Delete"};
                    AlertDialog.Builder builder = new AlertDialog.Builder(context);
                    builder.setTitle("Choose an action " );
                    builder.setItems(item, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            if (i==0){
                                dialogUpdate(getPosition());
                            }else {
                                removeProduct(getPosition());
                            }
                        }
                    });
                    builder.show();
                    return true;
                }
            });
        }


//        @Override
//        public void onCreateContextMenu(ContextMenu contextMenu, View view, ContextMenu.ContextMenuInfo contextMenuInfo) {
//            contextMenu.setHeaderTitle("Select  the action");
//            contextMenu.add(0,view.getId(),0,"UPDATE");
//            contextMenu.add(0,view.getId(),0,"DELETE");
//        }
    }

    private void removeProduct(int position) {
        Product product = productList.get(position);
        String id = product.getId();
        DatabaseHelper databaseHelper = new DatabaseHelper(context);
        databaseHelper.deleteProduct(id);
        ((MainActivity)context).hasdatainlist();
        ((MainActivity)context).setTotalTextView();

    }

    private void dialogUpdate(int position) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View viewLayout = inflater.inflate(R.layout.dialog_addproduct,null);
        builder.setView(viewLayout);
        textViewId =viewLayout.findViewById(R.id.edtID);
        edtName =viewLayout.findViewById(R.id.edtName);
        edtPrice= viewLayout.findViewById(R.id.edtPrice);
        edtDes =viewLayout.findViewById(R.id.edtDescription);
        ImageView img= viewLayout.findViewById(R.id.selectImageviewUpdate);
        Button btn =viewLayout.findViewById(R.id.btnAddProduct);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String name = edtName.getText().toString();
                String priceString = edtPrice.getText().toString();
                String des = edtDes.getText().toString();
              if(TextUtils.isEmpty(name)) {
                    edtName.setError("Please input name");
                    return;
                }else if(TextUtils.isEmpty(des)) {
                  edtDes.setError("Please input description");
                    return;
                }
                int priceInteger = 0;
                if (TextUtils.isEmpty(priceString)) {
                    edtPrice.setError("Please input price");
                }else {
                    try {
                        priceInteger = Integer.parseInt(priceString);
                    }catch (NumberFormatException e){
                        edtPrice.setError("Please input is number");
                    }


                }
                ////update
                if (priceInteger>0) {
                    Product product = productList.get(position);
                    DatabaseHelper databaseHelper = new DatabaseHelper(context);
                    databaseHelper.updateProducts(new Product(product.getId(), name, priceInteger, des));
                    ((MainActivity) context).hasdatainlist();
                    ((MainActivity)context).setTotalTextView();
                    alertDialog.dismiss();
                }else {
                    edtPrice.setError("Please input again");

                }
            }
        });
         alertDialog = builder.create();
        alertDialog.show();
    }

    public void setFilter(List<Product>newList){
        productList = new ArrayList<>();
        productList.addAll(newList);
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/example/producttest/adapter/ProductAdapter.kt
package com.example.producttest.adapter

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.example.producttest.R
import com.example.producttest.activity.MainActivity
import com.example.producttest.activity.ProductDetailActivity
import com.example.producttest.adapter.ProductAdapter.ViewHolderProduct
import com.example.producttest.database.DatabaseHelper
import com.example.producttest.fragments.ProductListFragment
import com.example.producttest.model.Product
import java.text.NumberFormat
import java.util.*


class ProductAdapter : RecyclerView.Adapter<ViewHolderProduct> {
    var context: Context
    var productList: MutableList<Product>
    private val PICK_IMAGE_REQUEST = 71
    var edtName: EditText? = null
    var edtPrice: EditText? = null
    var edtDes: EditText? = null
    var edtCount: EditText? = null
    var textViewId: TextView? = null
    var alertDialog: AlertDialog? = null
    var productListFragment: ProductListFragment? = null
    var mainActivity : MainActivity? = null

    constructor(context: Context, productList: MutableList<Product>) {
        this.context = context
        this.productList = productList
    }

    constructor(context: Context, productList: MutableList<Product>, productListFragment: ProductListFragment?) {
        this.context = context
        this.productList = productList
        this.productListFragment = productListFragment
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderProduct {
        val view = LayoutInflater.from(context).inflate(R.layout.item_product, parent, false)
        return ViewHolderProduct(view)
    }

    override fun onBindViewHolder(holder: ViewHolderProduct, position: Int) {
        val product = productList[position]
        //        holder.txtId.setText("Id "+product.getId());
        holder.txtName.text = product.name
        val locale = Locale("vi", "VN")
        val fmt = NumberFormat.getCurrencyInstance(locale)
        holder.txtPrice.text = fmt.format(product.price.toLong())
        holder.txtCount.text = product.count.toString() + ""
        holder.txtDes.text = product.des
        //
//        byte [] imageProduct = product.getImage();
//        Bitmap bitmap = BitmapFactory.decodeByteArray(imageProduct,0,imageProduct.length);
//        holder.imgProduct.setImageBitmap(bitmap);
        holder.imgDelete.setOnClickListener {
            val builder = androidx.appcompat.app.AlertDialog.Builder(context)
            builder.setTitle("Delete Cart")
            builder.setMessage("Do you want to delete: " + productList[position].name + "")
            builder.setPositiveButton("Yes") { dialogInterface, positions ->
                removeProduct(position)
                //                        ((MainActivity)context).setTotalTextView();
//                productListFragment!!.setTotalTextView()
                (context as MainActivity).setTotalTextView()
            }
            builder.setNegativeButton("No") { dialogInterface, i -> }
            builder.show()
        }

    }

    override fun getItemCount(): Int {
        return productList.size
    }

    private fun totalProduct(): Int {
        var total = 0
        for (product in productList) total += product.price
        Log.d("sd", total.toString() + "")
        return total
    }

    inner class ViewHolderProduct(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var txtName: TextView
        var txtPrice: TextView
        var txtDes: TextView
        var txtId: TextView? = null
        var txtCount: TextView
//        var imgProduct: ImageView
        var imgDelete: ImageView
        var img_Productss: ImageView
        private val linearLayoutDelete: LinearLayout? = null

        init {
            img_Productss = itemView.findViewById(R.id.img_Productss)
            txtName = itemView.findViewById(R.id.txtName)
            txtPrice = itemView.findViewById(R.id.txtPrice)
            txtDes = itemView.findViewById(R.id.txtDes)
//            imgProduct = itemView.findViewById(R.id.imgProduct)
            imgDelete = itemView.findViewById(R.id.delete_cart)
            txtCount = itemView.findViewById(R.id.txtCount)
            itemView.setOnClickListener { //
                val intent = Intent(context, ProductDetailActivity::class.java)
                intent.putExtra("ProductKey", productList[position])
                context.startActivity(intent)
                Log.d("ASDAS", "SADAS")
            }
            ///
            itemView.setOnLongClickListener {
                val item = arrayOf<CharSequence>("Update", "Delete")
                val builder = AlertDialog.Builder(context)
                builder.setTitle("Choose an action ")
                builder.setItems(item) { dialogInterface, i ->
                    if (i == 0) {
                        dialogUpdate(position)
                    } else {
                        removeProduct(position)
                    }
                }
                builder.show()
                true
            }
        }

    }

    private fun removeProduct(position: Int) {
        val product = productList[position]
        val id = product.id
        val databaseHelper = DatabaseHelper(context)
        databaseHelper.deleteProduct(id!!)
//        mainActivity!!.hasdatainlist();
//        mainActivity!!.setTotalTextView();
//        productListFragment!!.hasdatainlist()
//        productListFragment!!.setTotalTextView()
        (context as MainActivity).hasdatainlist()
        (context as MainActivity).setTotalTextView()
    }

    private fun dialogUpdate(position: Int) {
        val builder = AlertDialog.Builder(context)
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val viewLayout = inflater.inflate(R.layout.dialog_addproduct, null)
        builder.setView(viewLayout)
        textViewId = viewLayout.findViewById(R.id.edtID)
        edtName = viewLayout.findViewById(R.id.edtName)
        edtPrice = viewLayout.findViewById(R.id.edtPrice)
        edtCount = viewLayout.findViewById(R.id.edtCount)
        edtDes = viewLayout.findViewById(R.id.edtDescription)
        val img = viewLayout.findViewById<ImageView>(R.id.selectImageviewUpdate)
        val btn = viewLayout.findViewById<Button>(R.id.btnAddProduct)
        btn.setOnClickListener(View.OnClickListener {
            val products = productList[position]
            edtName!!.setText(products.name)
            val name = edtName!!.text.toString()
            val priceString = edtPrice!!.text.toString()
            val count = edtCount!!.text.toString().trim { it <= ' ' }
            val des = edtDes!!.text.toString()
            if (TextUtils.isEmpty(name)) {
                edtName!!.error = "Please input name"
                return@OnClickListener
            } else if (TextUtils.isEmpty(des)) {
                edtDes!!.error = "Please input description"
                return@OnClickListener
            }
            var priceInteger = 0
            if (TextUtils.isEmpty(priceString)) {
                edtPrice!!.error = "Please input price"
            } else {
                try {
                    priceInteger = priceString.toInt()
                } catch (e: NumberFormatException) {
                    edtPrice!!.error = "Please input is number"
                }
            }
            var countss = 0
            if (TextUtils.isEmpty(count)) {
                edtCount!!.error = "Please input count"
            } else {
                try {
                    countss = count.toInt()
                } catch (e: NumberFormatException) {
                }
            }
            ////update
            if (priceInteger > 0 && countss > 0) {
                val product = productList[position]
                val databaseHelper = DatabaseHelper(context)
                databaseHelper.updateProducts(Product(product.id, name, priceInteger, countss, des))
                //                    ((MainActivity) context).hasdatainlist();
//                    ((MainActivity)context).setTotalTextView();
                productListFragment!!.hasdatainlist()
                productListFragment!!.setTotalTextView()
                alertDialog!!.dismiss()
            } else {
                Toast.makeText(context, "Please input again", Toast.LENGTH_SHORT).show()
            }
        })
        alertDialog = builder.create()
        alertDialog!!.show()
    }

    fun setFilter(newList: List<Product>) {
        productList = ArrayList()
        productList!!.addAll(newList)
        notifyDataSetChanged()
    }
}